napari/components/_layer_slicer.py

```diff
@@ -209,8 +209,9 @@ def shutdown(self) -> None:
         # Replace with cancel_futures=True in shutdown when we drop support
         # for Python 3.8
         with self._lock_layers_to_task:
-            for task in self._layers_to_task.values():
-                task.cancel()
+            tasks = tuple(self._layers_to_task.values())
+        for task in tasks:
+            task.cancel()
```

Comment:
I think you've added this due to the ci failure, so I understand the *why*. 

It seems like if you move the loop over `tasks` out from inside the lock `with self._lock_layers_to_task:` (which I understand is required), then we should be able to just do away with the lock altogether, right?

If this loop is outside of the `with self._lock_layers_to_task:`

Reply:
My brain is a little fuzzy today, but here's my thought process.

- The error was the classic removing while iterating.
- Removing a future only occurs in the done callback.
- The done callback either gets executed when a pending future is cancelled or when a future completes.
- If a pending future is cancelled, the done callback is executed on the calling thread (in this case, the main thread). That's what caused the CI error here - there was only one thread involved and it was just the simple removing while iterating error. Holding the lock doesn't fix this case since it is reentrant.
- If a future completes, the done callback is executed on the executor's thread. Holding the lock prevents the done callback from removing here.

Does that make sense? And does the current solution seem good enough?

It might be worth adding separate tests for these two cases, especially since if/when we move to Python 3.9+, I'd want to check that `cancel_futures=True` does what we want.

Reply:
You are correct, while `tuple(self._layers_to_task.values())` is fast, its not instantaneous and therefore does need the lock.